Src/Base/AMReX_GpuFuse.cpp

```diff
@@ -0,0 +1,241 @@
+                }
+            } else {
+                int ncells = bx.numPts();
+                if (icell < ncells) {
+                    const auto len = amrex::length(bx);
+                    const auto lo  = amrex::lbound(bx);
+                    int k =  icell /   (len.x*len.y);
+                    int j = (icell - k*(len.x*len.y)) /   len.x;
+                    int i = (icell - k*(len.x*len.y)) - j*len.x;
+                    i += lo.x;
+                    j += lo.y;
+                    k += lo.z;
+                    if (helper.m_N == 0) {
+                        helper.m_fp.L3D(lambda_object,i,j,k);
+                    } else {
+                        for (int n = 0; n < helper.m_N; ++n) {
+                            helper.m_fp.L4D(lambda_object,i,j,k,n);
+                        }
+                    }
+                }
+            }
+        });
+        Gpu::synchronize();
+        The_Pinned_Arena()->free(nwarps);
+        The_Pinned_Arena()->free(h_buffer);
+        The_Device_Arena()->free(d_buffer);
+
+        for (int i = 0; i < nlambdas; ++i) {
+            char* p = m_lambda_buf + m_helper_buf[i].m_offset;
+            m_dtor_buf[i](p);
+            m_helper_buf[i].~FuseHelper();
+        }
+        m_dtor_buf.clear();
+        m_nbytes_used_lambda_buf = 0;
+        m_nlambdas = 0;
+    }
+}
+
+void
+Fuser::resize_lambda_buf ()
+{
+    m_nbytes_lambda_buf += m_nbytes_lambda_buf/2;
+    auto p = (char*)The_Pinned_Arena()->alloc(m_nbytes_lambda_buf);
+    std::memcpy(p, m_lambda_buf, m_nbytes_used_lambda_buf);
+    The_Pinned_Arena()->free(m_lambda_buf);
+    m_lambda_buf = p;
+}
+
+void
+Fuser::resize_helper_buf ()
+{
+    m_nhelpers_buf += m_nhelpers_buf/2;
+    auto p = (FuseHelper*)The_Pinned_Arena()->alloc(m_nhelpers_buf*sizeof(FuseHelper));
+    for (int i = 0; i < m_nlambdas; ++i) {
+        new (p+i) FuseHelper(m_helper_buf[i]);
+        (m_helper_buf+i)->~FuseHelper();
+    }
+    The_Pinned_Arena()->free(m_helper_buf);
+    m_helper_buf = p;
+}
+
+Fuser&
+Fuser::getInstance ()
+{
+    if (m_instance == nullptr) {
+        m_instance.reset(new Fuser());
+    }
+    return *m_instance;
+}
+
+void
+Fuser::Initialize ()
+{
+    ParmParse pp("amrex");
+    pp.query("gpu_fuse_size_threshold", s_fuse_size_threshold);
+    pp.query("gpu_fuse_numkernels_threshold", s_fuse_numkernels_threshold);
+
+    amrex::ExecOnFinalize(Fuser::Finalize);
+}
+
+void
+Fuser::Finalize ()
+{
+    m_instance.reset();
+}
+
+Long getFuseSizeThreshold () { return s_fuse_size_threshold; }
+
+Long
+setFuseSizeThreshold (Long new_threshold)
+{
+    Long old = s_fuse_size_threshold;
+    s_fuse_size_threshold = new_threshold;
+    return old;
+}
+
+int getFuseNumKernelsThreshold () { return s_fuse_numkernels_threshold; }
+
+int
+setFuseNumKernelsThreshold (int new_threshold)
+{
+    int old = s_fuse_numkernels_threshold;
+    s_fuse_numkernels_threshold = new_threshold;
+    return old;
+}
+
```

Comment:
I assume these get and set Threshold functions, plus the FuseSafeGuard are for an up-and-coming integration with the MFIter loop (add kernel launches that fit the parameters and launch in the destructor).

For users to be able to control the launches themselves, access to the current contents of the Fused Launch would be needed. (Number of launches awaiting launch in the fused region + some size metric, such as total number of points/threads in the launch thus far?)

Reply:
We can discuss in the later PR on integrating fuse into MFIter.

Reply:
Thanks for your comments.  But I want this PR to be focused on whether the approach of sending user's lambdas to gpu ourselves and calling them through function pointers is implemented correctly.

Reply:
Oh, that was my first pass through this. It looks solid to me. I learned some things too. (e.g. I don't think I've ever seen std::decay in action before).  :)
This might be going in a direction that there aren't a lot of experts for (except, maybe the RAJA developers who are doing something similar), so we may only really be able to determine if there's a problem (some edge case in the standard or compiler bugs) by solidly testing it.

Reply:
We use `std::decay` a lot in `AMReX_EB2_IF_*.H` and `amrex::GpuTuple`.  :)